src/kubernetes/reflector.rs

```diff
@@ -98,46 +111,40 @@ where
             };
 
             pin_mut!(stream);
-            while let Some(response) = stream.next().await {
-                // Any streaming error means the protocol is in an unxpected
-                // state. This is considered a fatal error, do not attempt
-                // to retry and just quit.
-                let response = response.map_err(|source| Error::Streaming { source })?;
-
-                // Unpack the event.
-                let event = match response {
-                    WatchResponse::Ok(event) => event,
-                    WatchResponse::Other(_) => {
-                        // Even though we could parse the response, we didn't
-                        // get the data we expected on the wire.
-                        // According to the rules, we just ignore the unknown
-                        // responses. This may be a newly added piece of data
-                        // our code doesn't know of.
-                        // TODO: add more details on the data here if we
-                        // encounter these messages in practice.
-                        warn!(message = "got unexpected data in the watch response");
-                        continue;
-                    }
-                };
-
-                // Prepare a resource version candidate so we can update (aka
-                // commit) it later.
-                let resource_version_candidate =
-                    match resource_version::Candidate::from_watch_event(&event) {
-                        Some(val) => val,
-                        None => {
-                            // This event doesn't have a resource version, this means
-                            // it's not something we care about.
+            loop {
+                // Obtain an value from the watch stream.
+                let val = if let Some(ref mut delayed_delete) = self.delayed_delete {
+                    // If delayed delete is requested, we perform the delayed
+                    // deletions concurrently with reading items from the watch
+                    // responses stream.
+                    let delayed_delete_deadline = delayed_delete.next_deadline();
+                    select! {
+                        // If we get a delayted delete deadline - process the
+                        // delayed deletes and restart the loop.
+                        _ = async { delay_until(delayed_delete_deadline.unwrap().into()).await }, if delayed_delete_deadline.is_some() => {
```

Comment:
This is good, but we can be as lazy with checking the deadline as we want so we could simplify this into a simple `delayed_delete.perform()` called every time after we get an event. It would make this more readable, like you had it before. Maybe less efficient, depending on how `Instant::new` is expensive. I'll leave it to you to decide.

Reply:
The problem is we won't be throwing away the deleted items longer than the timeout that way. It might not a serious problem, but this way the logic is cleaner (easier to reason about and test), so I'd rather keep this behavior.